cli/sessionstore/sessionstore.go

```diff
@@ -0,0 +1,166 @@
+}
+
+func (operatingSystemKeyring) Delete(service, user string) error {
+	return keyring.Delete(service, user)
+}
+
+const (
+	servicePrefix  = "coder-cli:"
+	keyringAccount = "session"
+)
+
+// Location represents where a session token is stored.
+type Location string
+
+const (
+	LocationNone    Location = "none"
+	LocationFile    Location = "file"
+	LocationKeyring Location = "keyring"
+)
+
+func serviceName(u *url.URL) string {
+	if u == nil || u.Host == "" {
+		return servicePrefix + "default"
+	}
+	host := strings.TrimSpace(strings.ToLower(u.Host))
+	return servicePrefix + host
+}
+
+// KeyringWithFallback is a Backend that prefers keyring storage and falls back to file
+// storage if the operating system keyring is unavailable. Happy path usage of this
+// type should start with NewKeyringWithFileFallback.
+type KeyringWithFallback struct {
+	keyringProvider keyringProvider
+}
+
+func NewKeyringWithFileFallback() KeyringWithFallback {
+	return NewKeyringProviderWithFileFallback(operatingSystemKeyring{})
+}
+
+func NewKeyringProviderWithFileFallback(provider keyringProvider) KeyringWithFallback {
+	return KeyringWithFallback{keyringProvider: provider}
+}
+
+// Read prefers reading the token from the operating system keyring and falls back
+// to file storage if the operating system keyring is unavailable.
+func (o KeyringWithFallback) Read(conf config.Root, serverURL *url.URL) (string, Location, error) {
+	svc := serviceName(serverURL)
+	tok, err := o.keyringProvider.Get(svc, keyringAccount)
+	if err == nil && tok != "" {
+		return tok, LocationKeyring, nil
+	}
+	// Fallback to file storage.
+	return File{}.Read(conf, serverURL)
+}
+
+// Write prefers storing the token in the operating system keyring and falls back
+// to file storage if the keyring operation fails.
+func (o KeyringWithFallback) Write(conf config.Root, serverURL *url.URL, token string) (Location, error) {
+	svc := serviceName(serverURL)
+	err := o.keyringProvider.Set(svc, keyringAccount, token)
+	if err == nil {
+		// Best effort: remove plaintext file if it exists.
+		_ = conf.Session().Delete()
+		return LocationKeyring, nil
+	}
+	// Fallback to file storage.
+	return File{}.Write(conf, serverURL, token)
+}
+
+// Delete removes any stored session token from both the keyring and file.
+func (o KeyringWithFallback) Delete(conf config.Root, serverURL *url.URL) error {
+	svc := serviceName(serverURL)
+
+	keyringErr := o.keyringProvider.Delete(svc, keyringAccount)
+	if keyringErr != nil {
+		if errors.Is(keyringErr, keyring.ErrNotFound) {
+			// Make the error handling for keyring/file backends uniform for the caller.
+			keyringErr = os.ErrNotExist
+		}
+	}
+
+	fileErr := File{}.Delete(conf, serverURL)
+	switch {
+	case fileErr != nil && keyringErr == nil:
+		// Happy path when using the keyring. We could drill down into fileErr, but for
+		// simplicity we will assume the error is because use of the keyring means the
+		// file didn't exist.
```

Comment:
No need to assume the file did not exist right?  We can check `os.ErrNotExist` and be sure.

Maybe we do something like this, to ensure all valid errors are reported:

```go
keyringErr := o.keyringProvider.Delete(svc, keyringAccount)
if errors.Is(keyringErr, keyring.ErrNotFound) {
	keyringErr = nil
}

fileErr := File{}.Delete(conf, serverURL)
if os.IsNotExist(fileErr) {
  fileErr = nil
}

return errors.Join(keyringErr, fileErr)
```